readthedocs/rtd_tests/tests/test_build_forms.py

```diff
@@ -43,10 +42,135 @@ def test_default_version_is_inactive(self):
 
         form = VersionForm(
             {
+                'slug': version.slug,
                 'active': False,
                 'privacy_level': PRIVATE,
             },
             instance=version,
         )
         self.assertFalse(form.is_valid())
         self.assertIn('active', form.errors)
+
+    def test_change_slug(self):
+        version = get(
+            Version,
+            project=self.project,
+            active=True,
+            slug='slug',
+        )
+
+        test_slugs = (
+            ('anotherslug', 'anotherslug'),
+            ('another_slug', 'another_slug'),
+            ('Another/Slug', 'another-slug'),
+        )
+        for slug, expected in test_slugs:
+            form = VersionForm(
+                {
+                    'slug': slug,
+                    'active': True,
+                    'privacy_level': PUBLIC,
+                },
+                instance=version,
+            )
+            self.assertTrue(form.is_valid())
+            self.assertEqual(version.slug, expected)
+
+    def test_change_slug_wrong_valud(self):
+        version = get(
+            Version,
+            project=self.project,
+            active=True,
+            slug='slug',
+        )
+
+        test_slugs = (
+            '',
+            '???//',
+            'long' * 100,
+        )
+        for slug in test_slugs:
+            form = VersionForm(
+                {
+                    'slug': slug,
+                    'active': True,
+                    'privacy_level': PUBLIC,
+                },
+                instance=version,
+            )
+            self.assertFalse(form.is_valid())
+            self.assertIn('slug', form.errors)
+
+    def test_change_slug_already_in_use(self):
+        version_one = get(
+            Version,
+            project=self.project,
+            active=True,
+            slug='one',
+        )
+        version_two = get(
+            Version,
+            project=self.project,
+            active=True,
+            slug='two',
+        )
+
+        form = VersionForm(
+            {
+                'slug': 'two',
+                'active': True,
+                'privacy_level': PUBLIC,
+            },
+            instance=version_one,
+        )
+        self.assertFalse(form.is_valid())
+        self.assertIn('slug', form.errors)
+
+
+    def test_cant_change_slug_machine_created_versions(self):
+        version = get(
+            Version,
+            project=self.project,
+            slug='machine',
+            active=True,
+            machine=True,
+        )
+        latest_version = self.project.versions.get(
+            slug='latest',
+        )
+
+        # Form works
+        form = VersionForm(
+            {
+                'slug': 'latest',
+                'active': False,
+                'privacy_level': PUBLIC,
+            },
+            instance=version,
+        )
+        self.assertTrue(form.is_valid())
```

Comment:
Why the form is valid if the version has `machine=True`?

Reply:
Django doesn't raise an exception if the value is changed (when disable=False), django just ignores the value, since users aren't suppose to change it.